Twitch Chat Events

mwseiden opened the pull request:
This is a mostly direct drop in from my Twitch IRC client code. Some issues with this code that need to be resolved:

No tests. I haven't even tested this as an MPF plugin manually. (in my code it's a scriptlet)
The IRC thread may need to be started later in the machine startup
The IRC thread should be able to be configured with machine variables. Currently it's config section based. Checking in passwords is bad.
The IRC thread should be stopped and restarted when those variables change.
The IRC thread is never shut down gracefully on exit
Is this the proper way to do threading in MPF?
It requires the irc library
Password token needs to be generated here: https://twitchapps.com/tmi/  the OAuth login should probably be integrated into MPF somehow

What it should do:
1. Log on to Twitch chat
2. Monitor chat for messages and post the `twitch_chat_message` event
3. Monitor chat for bit donations and post the `twitch_bit_donation` event
4. Monitor chat for subscriptions and post the `twitch_subscription` event

mwseiden committed:
Add twitch bot plug in

mwseiden committed:
Fix code

mwseiden committed:
Fix some variables

mwseiden committed:
Clean up code warnings

mwseiden committed:
Fix doc strings

mwseiden committed:
Fix more warnings

mwseiden committed:
Change chat event

mwseiden committed:
Update machine var code

jabdoa2 committed:
add config and a very basic test

jabdoa2 committed:
add test

jabdoa2 commented:
I added a basic test and config loading. Please have a look.

I guess we could port this to AioSimpleIRCClient (https://github.com/jaraco/irc/blob/master/irc/client_aio.py). Sending events and setting variables from another thread is not super safe. Should work most of the time but bad things may also happen. With an asyncio IRC client the thread would not be needed. Otherwise we need to change the calls slightly later.

mwseiden commented:
I can confirm that I get incredibly rare crashes on chat events. I know very little about asyncio, I don't have a quick answer on porting it.

jabdoa2 committed:
prevent races with another thread posting events

jabdoa2 commented:
Pushed a fix with separate threads. This is slightly slower but there shouldn't be too many events anyway. If you got time feel free to extend the test a bit. Currently it is doing almost nothing except making sure that the plugin loads.

jabdoa2 committed:
add basic test

jabdoa2 committed:
fix import

Commit:
Enable plugin

Commit:
Fix initialization

Commit:
Use token with or without oauth: prefix

Commit:
Add more tests

mwseiden commented:
Enabled the plugin in mpfconfig.yaml, fixed some code errors, added a couple more tests. Chat messages were tested by actually connecting to Twitch IRC.

Commit:
Add event docs

mwseiden commented:
Documentation PR: https://github.com/missionpinball/mpf-docs/pull/326

Commit:
Allow ! and ? commands to be passed to MPF

Commit:
Change a status message

jabdoa2 commented on mpf/plugins/twitch/twitch_client.py:
```diff
@@ -26,7 +26,10 @@ def __init__(self, machine, username, password, channel, loop):
-        super().__init__([(server, port, 'oauth:' + password)], username, username)
+        super().__init__([(
+            server, port,
+            (password if password.lower().startswith('oauth:') else ('oauth:' + password))
```
we could also have two parameters in config if that helps to make this more clear

mwseiden replied:
Most people are going to have the `oauth:` prefix on their passwords since that is what the token generator produces. I handle both because I did not copy the prefix and other people are going to make the same mistake.

jabdoa2 replied:
ok makes sense

jabdoa2 committed:
make linter happy and prevent crash when irc is not installed

jabdoa2 commented:
1. Do you want to merge it as it is or do you want to convert it to machine variables for credentials first?
2. Shutting down the thread gracefully seems to be missing in the irc library for some reason. This will probably prevent MPF from properly shutting down, right?

Commit:
Allow credentials to come from machine variables

Commit:
Merge branch 'feature-twitch' of https://github.com/mwseiden/mpf into feature-twitch

Commit:
Fix errors

Commit:
Fix formatting

mwseiden commented:
1. Added an option to use machine vars instead of hard coding. I would love to get this merged in as soon as possible.
2. It shuts down just fine. I have not noticed any bad effects from not stopping the thread first.

jabdoa2 commented:
Awesome. Thanks!
